Add CSV support to plotBuildData

langers2k opened the pull request.

langers2k commented:
I'll add doc's if that looks a sensible was of dealing with inclusion flags.

daspilker commented:
Hm, PlotXMLSeriesContext does not use an enum for the node types, so PlotCSVSeriesContext should not use an enum either. And depending on the inclusion flag, addiontial XML element like colExclusionSet and strExclusionSet must be generated.

How about adding dedicated methods instead of the enum?

```groovy
includeColumns(String... columnNames)
excludeColumns(String... columnNames)
includeColumns(int.. columIndexes)
excludeColumns(int.. columIndexes)
```

langers2k commented:
Hadn't noticed the two additional XML elements.

I'd rather keep the string than expect the user to add each index on a separate line. This is because selecting via indexes the plugin still expects a string of comma delimited values and when migrating to job-dsl it's just copy/paste.

Happy with the idea of dedicated methods but I'm thinking:
```groovy
includeByName(String criteria)
excludeByName(String criteria)
includeByIndex(String criteria)
excludeByIndex(String criteria)
```

Each of which will set `String inclusionFlag` and  either `String colExclusionSet` or `String strExclusionSet`.

langers2k commented:
Removed the enum and moved to dedicated methods with the elements that I'd missed previously

daspilker commented on job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext.groovy:
```diff
@@ -390,6 +397,22 @@ class PublisherContext implements Context {
+                                        }
+                                        if ((data.inclusionFlag == 'INCLUDE_BY_COLUMN') ||
+                                            (data.inclusionFlag == 'EXCLUDE_BY_COLUMN')) {
+                                            colExclusionSet()
```
There is something missing. The indexes need to be integers and then the `colExclusionSet` is filled. That's why I suggested to use int arguments in this case...

```xml
<hudson.plugins.plot.CSVSeries>
  <file>test</file>
  <label/>
  <fileType>csv</fileType>
  <colExclusionSet>
    <int>1</int>
    <int>2</int>
    <int>6</int>
  </colExclusionSet>
  <inclusionFlag>EXCLUDE_BY_COLUMN</inclusionFlag>
  <exclusionValues>1,2,6</exclusionValues>
  <url>bar</url>
  <displayTableFlag>true</displayTableFlag>
</hudson.plugins.plot.CSVSeries>
```

daspilker commented on job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext.groovy:
```diff
@@ -350,7 +358,6 @@ class PublisherContext implements Context {
      *                 </hudson.plugins.plot.PropertiesSeries>
      *                 <hudson.plugins.plot.XMLSeries>
      *                     <file>XML</file>
-     *                     <label></label>
```
The label element is generated for all types of series. Lame, but we should emulate that to keep the diff small.

daspilker commented:
There are merge conflicts because I removed all XML examples. Please rebase your commits to master.

langers2k commented:
Rebased and flattened.

- Added label to xml files
- Corrected usage of colExclusionSet and strExclusionSet
- Moved to suggestion for adding inclusion/exclusion sets
- Added checks to ensure that names/indexes aren't mixed up
- Added checks to ensure that includes/excludes aren't mixed up

daspilker commented:
Great. Update the docs and I will take care of the rest.
